docs/public-api.md

```diff
@@ -0,0 +1,97 @@
+# SwiftNIO 2 Public API
+
+The SwiftNIO project (which includes all of `github.com/apple/swift-nio*`) aims to follow [SemVer 2.0.0](https://semver.org/spec/v2.0.0.html) which requires projects to declare a public API. This document along with the [API documentation](https://apple.github.io/swift-nio/) declare SwiftNIO's API.
+
+## What are acceptable uses of NIO's Public API
+
+### Type, methods and modules
+
+It is acceptable and expected to use any type and call any method from the `NIO*` modules unless the type or function name start with an underscore (`_`). For `init`s this includes the first parameter's name.
+
+##### What's the point of this restriction?
+
+If we prefix something with an underscore or put it into one of our internal modules, we can't commit to an API for it and don't expect users to ever use this. Also, there might be different underscored functions/types depending on the platform.
+  
+##### Examples
+
+ - ✅ `channel.close(promise: nil)`
+ - ❌ `channel._unsafe.flush0()`, underscored property
+ - ❌ `import CNIOAtomics`, module name doesn't start with NIO
+ - ❌ `ByteBuffer(_enableSuperSpecialAllocationMode: true)`, as the initialiser's first argument is underscored
+
+---
+
+### Conforming NIO Types to Protocols
+
+It is acceptable to conform NIO types to protocols _you control_, ie. a protocol in _your codebase_. It is not acceptable to conform any NIO types to protocols in the standard library or outside of your control.
+
+##### What's the point of this restriction?
+
+NIO in a later new minor version might conform its type to said protocol.
```

Comment:
We may want to say here that NIO may conform its type to said protocol, *or* the package that provides the protocol may conform the NIO type to *their* protocol.